File: 2019/python/day1.py
#!/usr/bin/env python3
import sys


def day1(input_file):
    with open(input_file, "r") as f:
        input = f.read()

    print("Day 1: The Tyranny of the Rocket Equation")
    print(f"\tPart 1: {__part1(input)}")
    print(f"\tPart 2: {__part2(input)}")

def __recurse_fuel(val):
    return 0 if val//3-2 <= 0 else val//3-2 + __recurse_fuel(val//3-2)


def __part1(input):
    input = [int(x) for x in input.splitlines()]

    fuel = 0
    for mod in input:
        fuel += mod//3-2

    return fuel


def __part2(input):
    input = [int(x) for x in input.splitlines()]

    fuel = 0
    for mod in input:
        fuel += __recurse_fuel(mod)
    
    return fuel


if __name__ == "__main__":
    day1(sys.argv[1])

File: 2019/javascript/day1.js
#!/usr/bin/env node

import fs from "fs";

const part1 = (input) => {
  const mods = input.split("\n").map((e) => parseInt(e));
  return mods.reduce((acc, cur) => acc + (Math.trunc(cur / 3) - 2), 0);
};

const recurseFuel = (val) =>
  Math.trunc(val / 3) - 2 <= 0
    ? 0
    : Math.trunc(val / 3) - 2 + recurseFuel(Math.trunc(val / 3) - 2);

const part2 = (input) => {
  const mods = input.split("\n").map((e) => parseInt(e));
  return mods.reduce((acc, cur) => acc + recurseFuel(cur), 0);
};

const day1 = (inputFile) => {
  const input = fs.readFileSync(inputFile, "utf8");

  console.log("Day 1: The Tyranny of the Rocket Equation");
  console.log(`\tPart 1: ${part1(input)}`);
  console.log(`\tPart 2: ${part2(input)}`);
};

export default day1;

try {
  day1(process.argv[2]);
} catch (e) {
  console.log(`${e}`);
}
